pkg/controller/direct/notebooks/instance_controller.go

```diff
@@ -0,0 +1,269 @@
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type InstanceAdapter struct {
+	id        *krm.InstanceIdentity
+	gcpClient *gcp.NotebookClient
+	desired   *krm.NotebookInstance
+	actual    *notebookspb.Instance
+}
+
+var _ directbase.Adapter = &InstanceAdapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return false means the object is not found. This triggers Adapter `Create` call.
+// Return a non-nil error requeues the requests.
+func (a *InstanceAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting Instance", "name", a.id)
+
+	req := &notebookspb.GetInstanceRequest{Name: a.id.String()}
+	instancepb, err := a.gcpClient.GetInstance(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting Instance %q: %w", a.id, err)
+	}
+
+	a.actual = instancepb
+	return true, nil
+}
+
+// Create creates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *InstanceAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating Instance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desired := a.desired.DeepCopy()
+	resource := NotebookInstanceSpec_ToProto(mapCtx, &desired.Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	req := &notebookspb.CreateInstanceRequest{
+		Parent:     a.id.Parent().String(),
+		InstanceId: a.id.ID(),
+		Instance:   resource,
+	}
+	op, err := a.gcpClient.CreateInstance(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating Instance %s: %w", a.id, err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("instance %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created Instance", "name", a.id)
+
+	status := &krm.NotebookInstanceStatus{}
+	status.ObservedState = NotebookInstanceObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.LazyPtr(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *InstanceAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating Instance", "name", a.id)
+	mapCtx := &direct.MapContext{}
+
+	desiredPb := NotebookInstanceSpec_ToProto(mapCtx, &a.desired.DeepCopy().Spec)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	paths, err := common.CompareProtoMessage(desiredPb, a.actual, common.BasicDiff)
+	if err != nil {
+		return err
+	}
+
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
```

Comment:
We still want to update the KRM status here. Assuming user is acquiring the GCP resource with the same spec, we should populate the observed state with "a.actual" and also add the "externalRef".